Triple::normalize: Use none as OS for XX-none-ABI

wzssyqa opened the pull request:
When parsing a 3-component triple, after we determine Arch and Env, if the middle component is "none", treat it as OS instead of Vendor.

See: https://discourse.llvm.org/t/rfc-baremetal-target-triple-normalization/78524
Fixes: #89582.

wzssyqa commented:
@peterwaller-arm Ohh, there is so many `-none-unknown-` in current code. I guess it may be widely used.
Do we really want to change all of them?

wzssyqa commented:
@llvm/pr-subscribers-backend-arm @llvm/pr-subscribers-backend-aarch64

sdesmalen-arm commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -1149,6 +1149,13 @@ std::string Triple::normalize(StringRef Str) {
+  // This is for some baremetal cases, such as "arm-none-elf".
+  if (Found[0] && !Found[1] && !Found[2] && Found[3] &&
+      Components[1].equals("none") && Components[2].empty())
+    std::swap(Components[1], Components[2]);
```
This change, along with the change in libcxx/utils/ci/run-buildbot, look like the only functional changes in this patch. Is it worth committing the non-functional changes (i.e. most of the tests) in a separate NFC patch?

wzssyqa replied:
Sure. we can. I think that we may need more update to `Triple::normalize`. Once we can decide how to normalize `aarch64-none-elf`, then I will split this PR.

DavidSpickett commented:
> Ohh, there is so many -none-unknown- in current code. I guess it may be widely used. Do we really want to change all of them?

In current test cases, which were likely generated using clang. So they are likely not relying on that exact form of the triple, it's just what clang put out at the time. So they don't prove anything either way, unless they fail after this change of course.

Thank you Peter for testing this PR, I've also asked another downstream team to test their bare metal builds with this. That'll be our best indication for now if this makes sense.

peterwaller-arm commented:
I've just posted an RFC to discourse about how to handle triple normalization: https://discourse.llvm.org/t/rfc-baremetal-target-triple-normalization/78524

wzssyqa commented:
> I've just posted an RFC to discourse about how to handle triple normalization: https://discourse.llvm.org/t/rfc-baremetal-target-triple-normalization/78524

We don't need to revert https://github.com/llvm/llvm-project/pull/89234
It is easy to fix `Triple::normalize` once you decide how to normalize your triples.
I will help to do so.

MaskRay commented:
The description is a bit difficult to parse. I'll suggest that using some grammar checker to fix the description.
But thanks for putting up the patch. I think it is desired.

https://discourse.llvm.org/t/rfc-baremetal-target-triple-normalization/78524/8

> The number of none-unknown test changes actually looks manageable.
The worst case is that we notice some backward compatibility story that we don’t account for today,
and we add temporary workarounds for clang --target=aarch64-none-elf to probe lib/aarch64-none-unknown-elf (old LLVM_ENABLE_PER_TARGET_RUNTIME_DIR hierarchy)
beside lib/aarch64-unknown-none-elf.

wzssyqa committed:
Triple::normalize: Use none as OS for XX-none-ABI

When parsing a 3-component triple, after we determine Arch and Env,
if the middle component is "none", treat it as OS instead of Vendor.

Fixes: #89582.

peterwaller-arm commented:
Thanks @wzssyqa again for implementing this and splitting bits out per Sander's suggestion. I propose we merge these changes tomorrow, on the understanding that each of the two patches (this and #90313) are passing tests. It looks to me like you're working towards that goal.

What I see in CI for this PR is `LLVM-Unit :: TargetParser/./TargetParserTests/TripleTest/Normalization` is currently failing.

Is it possible to have two patches such that they're both green? My expectation is that the majority of the triple changes in the source are unrelated to normalization and would work the same before and after the functional changes to triple normalization. (And that those tests which are impacted by changes to triple normalization should of course change in the same commit to keep things green).

wzssyqa committed:
fix test failures

wzssyqa commented:
@peterwaller-arm half-done. Do *NOT* merge it now. It has still some problem.
See `clang/test/Driver/baremetal-multilib.yaml`.

There is still some `none-unknown`.

wzssyqa commented:
@peterwaller-arm It seems OK now. Let's wait the result of CI.

MaskRay commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -1151,6 +1151,13 @@ std::string Triple::normalize(StringRef Str) {
+  // while if it is "none", we'd prefer to set OS.
+  // This is for some baremetal cases, such as "arm-none-elf".
+  if (Found[0] && !Found[1] && !Found[2] && Found[3] &&
+      Components[1].equals("none") && Components[2].empty())
```
Prefer `==` to `equals`.

`equals` is to be deprecated

MaskRay commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -1151,6 +1151,13 @@ std::string Triple::normalize(StringRef Str) {
     }
   }
 
+  // For 3-component triples, the middle component is used to set Vendor;
```
Suggested comment

If "none" is in the middle component in a three-component triple, treat it as
the OS (Components[2]) instead of the vendor (Components[1]).

MaskRay commented:
> When we parse 3-components triples, if the Arch and Env have been parsed successfully, we have to make a choice between Vendor and OS for the unrecoginzed component. Noramlly it is the middle one.
> 
> In the current code, Vendor is choosed, and then OS is fallbacked to unknown. It is OK for most cases. But if the unrecoginzed component is `none`, it is expected to be OS instead of Vendor.

Suggested:

When parsing a 3-component triple, after we determine Arch and Env, if the middle component is "none", treat it as OS instead of Vendor.

DavidSpickett commented:
I'm checking the picolib libcxx build right now so please hold off on landing this.

peterwaller-arm commented on libcxx/utils/ci/run-buildbot:
```diff
@@ -217,7 +217,7 @@ function test-armv7m-picolibc() {
 
     ${NINJA} -vC "${BUILD_DIR}/compiler-rt" install
-    mv "${BUILD_DIR}/install/lib/armv7m-none-unknown-eabi"/* "${BUILD_DIR}/install/lib"
+    mv "${BUILD_DIR}/install/lib/armv7m-unknown-none-eabi"/* "${BUILD_DIR}/install/lib"
```
This triple is 'used' by the version of the clang on the builder. Please remove this update, I'm in touch with @DavidSpickett who will make sure it gets updated at the right time.

DavidSpickett replied:
https://github.com/llvm/llvm-project/pull/90722

You can remove all changes to this file.

peterwaller-arm reviewed (changes requested):
Thanks for doing this and waiting for RFC comments.

I think this can be merged, the one thing which needs fixing is removing the change I've commented on so that the picolibc builder can succeed (it uses a version of clang on the builder, which @DavidSpickett will update at the right time).

DavidSpickett commented:
Also this change should be release noted with a tip that build scripts may want to change to asking clang for the triple instead.

peterwaller-arm reviewed (approved):
LGTM, thanks for supplying a quick fix and iterating this, and thanks for waiting for input on the RFC.

DavidSpickett commented:
libcxx is sorted, and I will write up the release note given that I suggested it and have been fixing some builds myself.

MaskRay commented:
#90734 provided a release note. 

The `armv7m-none-eabi` example can probably be copied as an example to improve the commit message/description.

DavidSpickett commented:
@wzssyqa thanks for working with us on this. Bare metal is always a wild card but I think in the end we'll all have cleaner builds because of these changes.
